File: src/problems/Problem30.java
package problems;

public class Problem30 {
	public static void main(String[] args) {
		System.out.println(solution());
	}

	public static int solution() {
		int sum = 0;
		for (int i = 2; i < 200000; i++) {
			if (i == sum(i))
				sum+=i;
		}
		return sum;
	}
	
	private static int sum(int i) {
		int sum = pow[i%10];
		while ((i/=10) != 0) {
			sum+=pow[i%10];
		}
		return sum;
	}

	static int[] pow = new int[10];
	
	static {
		for (int i = 0; i < pow.length; i++) {
			pow[i] = i*i*i*i*i;
		}
	}
}

File: src/main/scala/Problem30.scala
package scala

object Problem30 extends App {
  def sum(n: Long): Long = {
    if (n == 0) 0 else Math.pow(n % 10, 5).toLong + sum(n / 10)
  }
  println((2L to 1000000L).filter(x => x == sum(x)).sum)
}
